src/forms/form-fields/helpers/IconPickerDialog/IconPickerCustomTab.js

```diff
@@ -0,0 +1,210 @@
+import React, { Component } from 'react';
+import Button from 'd2-ui/lib/button/Button';
+import RaisedButton from 'material-ui/RaisedButton';
+import PropTypes from 'prop-types';
+import Divider from 'material-ui/Divider';
+import TextField from 'material-ui/TextField/TextField';
+import { fileToBase64 } from './fileToBase64.js';
+import { uploadIcon } from './uploadIcon.js';
+import ErrorMessage from 'd2-ui/lib/messages/ErrorMessage.component';
+import classes from 'classnames';
+
+const styles = {
+    selectFileButton: {
+        backgroundColor: '#ff9800',
+        color: '#fff',
+        textAlign: 'center',
+        position: 'relative',
+        minWidth: 129,
+        height: 36,
+        lineHeight: 2.5,
+        boxShadow: '0 1px 6px rgba(0,0,0,0.12),0 1px 4px rgba(0,0,0,0.12)',
+        cursor: 'pointer',
+        textTransform: 'uppercase',
+    },
+    iconFileInput: {
+        display: 'none',
+    },
+};
+
+export class IconPickerCustomTab extends Component {
+    constructor(props, context) {
+        super(props, context);
+
+        this.t = context.d2.i18n.getTranslation.bind(context.d2.i18n);
+        this.d2 = context.d2;
+        this.state = {
+            iconFile: null,
+            iconFileBase64: null,
+            iconMetadata: {
+                key: '',
+                description: '',
+                keywords: [],
+            },
+            uploading: false,
+            uploadError: null,
+        };
+    }
+
+    handleIconMetadataChange = event => {
+        const value = event.target.value && event.target.value.trim();
+        const name = event.target.name;
+
+        this.setState(prevState => ({
+            iconMetadata: {
+                ...prevState.iconMetadata,
+                [name]: value,
+            },
+        }));
+    };
+
+    handleIconKeywordsChange = event => {
+        const keywords = event.target.value
+            .split(',')
+            .map(kw => kw.trim())
+            .filter(kw => !!kw);
+        this.setState(prevState => ({
+            iconMetadata: {
+                ...prevState.iconMetadata,
+                keywords,
+            },
+        }));
+    };
+
+    handleFileUpload = () => {
+        const iconData = this.state.iconMetadata;
+        this.setState({ uploading: true });
+        uploadIcon(this.state.iconFile, iconData)
+            .then(() => {
+                if (typeof this.props.onIconUpload === 'function') {
+                    this.props.onIconUpload(iconData.key);
+                }
+                this.setState({ uploading: false, uploadError: null });
+            })
+            .catch(e => this.setState({ uploading: false, uploadError: e }));
+    };
+
+    handleFileChange = async event => {
+        const file = event.target.files[0];
+        if (!file) {
```

Comment:
it would probably be good to limit `file.size` to something reasonable (in the future)

Reply:
There's server side validation for this: 
![image](https://github.com/dhis2/maintenance-app/assets/13852438/85f6b1f1-3ae9-443c-86f1-11bae3f34ddd)

Want me to add that to client as well? I do think this could be changed in server config, so maybe enough to depend on the server for this?